pilot/pkg/config/mcp/controller_test.go

```diff
@@ -0,0 +1,393 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package coredatamodel_test
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	google_protobuf "github.com/gogo/protobuf/types"
+	"github.com/onsi/gomega"
+	mcp "istio.io/api/mcp/v1alpha1"
+	networking "istio.io/api/networking/v1alpha3"
+	coredatamodel "istio.io/istio/pilot/pkg/config/mcp"
+	"istio.io/istio/pilot/pkg/config/mcp/fakes"
+	"istio.io/istio/pilot/pkg/model"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+)
+
+func TestRegisterEventHandler(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	logger := &fakes.Logger{}
+	controller := coredatamodel.NewController(logger)
+
+	var (
+		registeredEvent     int
+		registeredEventLock sync.Mutex
+	)
+
+	stop := make(chan struct{})
+	go controller.Run(stop)
+	defer func() {
+		stop <- struct{}{}
+	}()
+
+	controller.RegisterEventHandler("virtual-service", func(model.Config, model.Event) {
+		registeredEventLock.Lock()
+		registeredEvent++
+		registeredEventLock.Unlock()
+	})
+
+	g.Eventually(func() int {
+		registeredEventLock.Lock()
+		defer registeredEventLock.Unlock()
+		return registeredEvent
+	}).Should(gomega.Equal(1))
+}
+
+func TestConfigDescriptor(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	logger := &fakes.Logger{}
+	controller := coredatamodel.NewController(logger)
+
+	descriptors := controller.ConfigDescriptor()
+	g.Expect(descriptors).To(gomega.Equal(model.IstioConfigTypes))
+}
+
+func TestGetInvalidType(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	logger := &fakes.Logger{}
```

Comment:
Please don't fake logger in pilot. 
There is no need - let it log, all tests we have do this, no need for MCP to be different.

Reply:
👍